Implement support for actions workflow runs

ayzu opened the pull request:
Relates to: #1399.

gmlewis commented:
Thank you, @ayzu . It looks like you need to merge master into your PR because it is modifying files it shouldn't which might mean that you based your work on another PR that got changed during a code review. ... If I'm making any sense at all.

I can be more detailed if this doesn't make sense... But that should fix the broken tests.

ayzu commented:
> Thank you, @ayzu . It looks like you need to merge master into your PR because it is modifying files it shouldn't which might mean that you based your work on another PR that got changed during a code review. ... If I'm making any sense at all.
> 
> I can be more detailed if this doesn't make sense... But that should fix the broken tests.

Thank you, @gmlewis 

I didn't run `go-generate` :) By the way, what do you think of adding a Makefile like:

```
do:
	go generate ./...
	go test ./...
	go vet ./...
```

gmlewis commented:
> I didn't run go-generate :) By the way, what do you think of adding a Makefile like:

:smile: Funny you should mention it. I've had this bash script in my directory for a long time:

```bash
$ cat testit.sh 
#!/bin/bash -ex
GO111MODULE=on
go generate ./...
go test ./...
go vet ./...
```

I could probably get rid of that environment variable now.  :smile:

Adding a `Makefile` to Go projects just brings back memories of waiting for C++ compilations, so I'd rather not.  :smile:

gmlewis commented on github/actions_workflow_runs.go:
```diff
@@ -0,0 +1,197 @@
+	HeadRepository *Repository       `json:"head_repository,omitempty"`
+}
+
+// Runs represents a slice of repository action workflow run.
```
Let's please rename this to `WorkflowRuns`.

gmlewis commented on github/actions_workflow_runs.go:
```diff
@@ -0,0 +1,197 @@
+	"net/url"
+)
+
+// PersonWorkflowRun represents a person object as returned in actions workflow jobs endpoint response.
```
Let's please remove this struct entirely. See below.

gmlewis commented on github/actions_workflow_runs.go:
```diff
@@ -0,0 +1,197 @@
+	Email *string `json:"email,omitempty"`
+}
+
+// CommitWorkflowRun represents a commit object as returned in actions workflow jobs endpoint response.
```
Let's please remove this struct entirely. See below.

gmlewis commented on github/actions_workflow_runs.go:
```diff
@@ -0,0 +1,197 @@
+	ArtifactsURL   *string           `json:"artifacts_url,omitempty"`
+	CancelURL      *string           `json:"cancel_url,omitempty"`
+	RerunURL       *string           `json:"rerun_url,omitempty"`
+	HeadCommit     CommitWorkflowRun `json:"head_commit,omitempty"`
```
I see in this repo that we have a very similar struct that appears to serve the same purpose that I think might be good to reuse here: `PushEventCommit` in `event_types.go`, although some might argue with me that `PushEventCommit` is strictly to be used for events.

I would argue that the `HeadCommit` is how that struct is being used, and maybe it should be renamed to `HeadCommit` and then used in both places to represent the same thing.

Either way, whether `PushEventCommit` is renamed to `HeadCommit` in this PR or not, this instance of `HeadCommit` should be a pointer to a struct, not an embedded struct.

As for the rename of `PushEventCommit`, I'll leave it up to you and the second reviewer of this PR, as I could go either way.

gmlewis commented on github/actions_workflow_runs.go:
```diff
@@ -0,0 +1,197 @@
+	NodeID         *string           `json:"node_id,omitempty"`
+	HeadBranch     *string           `json:"head_branch,omitempty"`
+	HeadSHA        *string           `json:"head_sha,omitempty"`
+	RunNumber      *int64            `json:"run_number,omitempty"`
```
I'm thinking `RunNumber` should be an `*int` here, not an `*int64`.

gmlewis commented on github/actions_workflow_runs.go:
```diff
@@ -0,0 +1,197 @@
+// Runs represents a slice of repository action workflow run.
+type Runs struct {
+	TotalCount *int           `json:"total_count,omitempty"`
+	Runs       []*WorkflowRun `json:"workflow_runs,omitempty"`
```
s/Runs/WorkflowRuns/
for consistency with the JSON parameter name.

gmlewis commented on github/actions_workflow_runs.go:
```diff
@@ -0,0 +1,197 @@
+// ListWorkflowRunsByFileName lists all workflow runs by workflow file name.
+//
+// GitHub API docs: https://developer.github.com/v3/actions/workflow_runs/#list-workflow-runs
+func (s *ActionsService) ListWorkflowRunsByFileName(ctx context.Context, owner, repo string, workflowFileName string, opts *ListWorkflowRunsOptions) (*Runs, *Response, error) {
```
`... owner, repo, workflowFileName string,...`

ayzu commented:
> This is looking great, @ayzu!
> Just a few tweaks, please, and then I think we'll be ready for a second review before merging.

Thank you for the review, @gmlewis 

I updated the code and renamed the structure to `HeadCommit`. It is great that we can reuse an existing type.

gmlewis commented on github/actions_workflow_runs_test.go:
```diff
@@ -31,11 +31,11 @@ func TestActionsService_ListWorkflowRunsByID(t *testing.T) {
-			{ID: Int64(399444496), RunNumber: Int64(29679449), CreatedAt: &Timestamp{time.Date(2019, time.January, 02, 15, 04, 05, 0, time.UTC)}, UpdatedAt: &Timestamp{time.Date(2020, time.January, 02, 15, 04, 05, 0, time.UTC)}},
-			{ID: Int64(399444497), RunNumber: Int64(29679449), CreatedAt: &Timestamp{time.Date(2019, time.January, 02, 15, 04, 05, 0, time.UTC)}, UpdatedAt: &Timestamp{time.Date(2020, time.January, 02, 15, 04, 05, 0, time.UTC)}},
+		WorkflowRuns: []*WorkflowRun{
+			{ID: Int64(399444496), RunNumber: Int(29679449), CreatedAt: &Timestamp{time.Date(2019, time.January, 02, 15, 04, 05, 0, time.UTC)}, UpdatedAt: &Timestamp{time.Date(2020, time.January, 02, 15, 04, 05, 0, time.UTC)}},
```
nit: for the `RunNumber` values, let's kick this down to maybe `296` or another relatively small number so that we don't give the impression that these are IDs from a quick glance.

gmlewis commented:
Thank you, @wesleimp!
Merging.
